### DISCUSSION TITLE
sensor.foxess_em_forecast exceed maximum size???

### DISCUSSION THREAD

**Discussion by C00K50N** [nathanmarlor/foxess_em#279] [General]
I love this integration!  It almost makes our system "set and forget".

I'm not sure it's working quite right at the moment though, and I'm trying to trace why.  For info, we use our Fox Cloud details within EM to handle the charging, and the inverter's own LAN port for reading data via StealthChesnut's Modbus integration (we're on Manager 1.57 and it still works fine - we skipped a few Manager versions so didn't have it drop out).  

The last few days our batteries have charged to full on our cheap rate, despite the forecast being good enough to only require a partial or no charge.  It seems to have reverted back to our manual settings of a force charge for the full duration of our cheap period. Last night I changed the manual setting to force charge for only an hour.  It did so, then used the battery for house load after that hour, rather than holding as it would when this integration is working.

The current Fox server updates / instability might be the cause but I'm not sure as I found something else in the logs...  Looking back, this warning below seems to have been coming up periodically for the last couple weeks.

So... is there anything I can do?  Do I need to clear database data in order to make space?  If so how do I go about it, and what implications would that have for longer term stats?  Is there a way to automatically clear or better compress older data within the HA database?  

2023-05-03 01:27:10.728 WARNING (Recorder) [homeassistant.components.recorder.db_schema] State attributes for sensor.foxess_em_forecast exceed maximum size of 16384 bytes. This can cause database performance issues; Attributes will not be stored
2023-05-03 01:27:11.842 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved

**Comment by nathanmarlor** [comment]
Thanks @C00K50N !

I'd strongly recommend setting up EM with a RS485 device - especially given the latest Fox Cloud instabilities. You should see failed calls in the HA logs at the time your off-peak period starts?

Regarding the log message - it's nothing to worry about, just add the following to your HA config:

```
recorder:
  exclude:
    entities:
      - sensor.foxess_em_raw_data
      - sensor.foxess_em_forecast
```

I'll move to use the proper HA storage options at some point which will eradicate this.

**Reply by C00K50N** [reply]
Thanks for the quick reply.  So the warning is nothing to worry about?

Apologies for repeating the same query in two places, but since you mention it...  Our direct LAN connection and StealthChesnut integration has been absolutely rock solid from Manager 1.44 and for the last couple months on 1.57 (we managed to skip the ones that cut it off but 1.57 works just the same as before).  

As a novice, the prospect of adding and setting up a further device is a daunting one!  With that in mind, are there any plans to work with the direct LAN connection for EM or your Modbus integration?  Or will it work already and I've got confused?!

**Reply by nathanmarlor** [reply]
It's warning you that the attribute size on the entity is too large to be stored, so will be skipped - this is fine and actually preferred to avoid bloating the HA database; adding the configuration above explicitly instructs HA to ignore it.

Unfortunately EM doesn't currently support direct LAN connections as the charge control registers haven't yet been mapped. There are plenty of guides stepping through how to setup and configure an RS485 device, plus lots of people willing to help across the community :)

**Reply by C00K50N** [reply]
Understood, thanks for the clarification.  Yes, the community is super-helpful!  Without it I would still be looking at the forecast each evening and manually working out how long to charge for overnight!  Not to mention getting frustrated at the non-live but supposedly live energy data!!

Thanks for all your help.  I'll add an RS485 device to my to-do list for when I'm feeling brave!

**Comment by C00K50N** [comment]
Hi @nathanmarlor, sorry to hassle you again...  while I will switch to an RS485 device at some point, for the time being I could do with carrying on just using the Cloud connection.  I updated this integration to v1.7.5 in the hope that would solve the issue but it hasn't - it's still charging to full each night based on my manual settings on the app / web.  

I've gone back through the config for EM, and when it gets to entering the Fox Cloud details I get an Error - check username and password.  I have checked and double-checked that my login details haven't changed, and the same details still let me log in to Fox Cloud on the web and app, just not it seems with this integration.  Can you offer any ideas?  Have Fox perhaps limited logins to only those from their own website while they're updating stuff this week?  Would they even be able to tell the difference?

![image](https://user-images.githubusercontent.com/102669305/236434469-e7a4239f-542a-4606-9b49-7d7b00f0c40e.png)

**Reply by nathanmarlor** [reply]
No problem! 

They wouldn't be able to tell the difference, we use the API the same as the JavaScript running in your browser.

Can you grab some logs please?

**Reply by C00K50N** [reply]
Aaargh... I've just restarted HA after adding the exclude code you suggested the other day, so I only have logs from when it came back up.  D'oh!

Looks like it was trying to log in, but I don't know if you'd expect it to tell me if it succeeded or not...

2023-05-05 11:32:09.250 DEBUG (MainThread) [custom_components.foxess_em.fox] Logging into Fox Cloud
2023-05-05 11:32:09.250 DEBUG (MainThread) [custom_components.foxess_em.fox] Issuing request to (https://www.foxesscloud.com/c/v0/user/login) with params: {'user': '**[my username]**', 'password': '**[long string of hex, not my actual password?]**'}

No errors as far as I can see.

**Reply by C00K50N** [reply]
I've just tried changing my password to see if it made a difference.  The Fox Cloud website let me straight back in with new details, but EM doesn't like my new password either.

**Reply by nathanmarlor** [reply]
Interesting - nothing at all after the last line? Perhaps Fox have changed the API after the last update, I'll do some digging! 

FYI this inability to login explains the full charges - the integration will always 'fail safe' and ensure your batteries are fully charged rather than depleted in the event of a Fox outage/error.

**Reply by C00K50N** [reply]
Yes, that fail safe is ideal, exactly what we need.  Just not as perfect as when the EM integration is working fully!

There were plenty more lines in the log, but no more [custom_components.foxess_em.fox] ones after these two as far as I can see...

Although as I've looked again just now, it seems my further attempts to log in may have triggered this warning:

2023-05-05 11:41:45.173 WARNING (MainThread) [custom_components.foxess_em.config_flow] 'NoneType' object is not subscriptable

Furthermore, after I tried to login with my new password - and also tried inputting my username all lowercase - each time I tried to log in it re-populated the username and password fields with my **old** details.  Is HA storing them in a cache somewhere I could clear out?  Or anything like that?

**Reply by nathanmarlor** [reply]
It'll store your password once it's been validated so it's not surprising that it's defaulting back to the old (last working) ones.

**Reply by C00K50N** [reply]
OK.  Hmm... not sure what else I can try here.  Let me know if you're able to find any changes they've made at their end which could be causing the problem.  Or if there's anyone else seeing the same.  Maybe I'm the only one still connecting via Fox Cloud?!

**Reply by nathanmarlor** [reply]
@C00K50N - Apologies for the delay on this one. It was a change to the Fox cloud as suspected - now fixed in https://github.com/nathanmarlor/foxess_em/releases/tag/v1.7.6

**Reply by C00K50N** [reply]
@nathanmarlor - That's great, thank you.  Just updated now, and looks like the login worked fine as previous.  Excellent!  It will be great to have it up and running again.  Thank you again.
